src/PackagePrivate/DatabaseTermIdsAcquirer.php

```diff
@@ -0,0 +1,420 @@
+<?php
+
+namespace Wikibase\TermStore\MediaWiki\PackagePrivate;
+
+use Wikimedia\Rdbms\IDatabase;
+
+class DatabaseTermIdsAcquirer implements TermIdsAcquirer {
+
+	const TABLE_TEXT = 'wbt_text';
+	const TABLE_TEXT_IN_LANG = 'wbt_text_in_lang';
+	const TABLE_TERM_IN_LANG = 'wbt_term_in_lang';
+
+	/**
+	 * @var Database $dbw
+	 */
+	private $dbw;
+
+	/**
+	 * @var Database $dbr
+	 */
+	private $dbr;
+
+	public function __construct(
+		IDatabase $dbw,
+		IDatabase $dbr,
+		TypeIdsAcquirer $typeIdsAcquirer
+	) {
+		$this->dbw = $dbw;
+		$this->dbr = $dbr;
+		$this->typeIdsAcquirer = $typeIdsAcquirer;
+	}
+
+	public function acquireTermIds( array $termsArray ): array {
+		$termsArray = $this->mapToTextIds( $termsArray );
+		$termsArray = $this->mapToTextInLangIds( $termsArray );
+		$termsArray = $this->mapToTypeIds( $termsArray );
+		return $this->mapToTermInLangIds( $termsArray );
+	}
+
+	/**
+	 * replace root keys containing type names in termsArray
+	 * with their respective ids in wbt_type table
+	 *
+	 * @param array $termsArray terms per type per language:
+	 *	[
+	 *		'type1' => [ ... ],
+	 *		'type2' => [ ... ],
+	 *		...
+	 *	]
+	 *
+	 * @return array
+	 *	[
+	 *		<typeId1> => [ ... ],
+	 *		<typeId2> => [ ... ],
+	 *		...
+	 *	]
+	 */
+	private function mapToTypeIds( array $termsArray ) {
+		$typeIds = $this->typeIdsAcquirer->acquireTypeIds( array_keys( $termsArray ) );
+
+		$termsArrayByTypeId = [];
+		foreach ( $typeIds as $type => $typeId ) {
+			$termsArrayByTypeId[$typeId] = $termsArray[$type];
+		}
+		return $termsArrayByTypeId;
+	}
+
+	/**
+	 * replace text at termsArray leaves with their ids in wbt_text table
+	 * and return resulting array
+	 *
+	 * @param array $termsArray terms per type per language:
+	 *	[
+	 *		'type' => [
+	 *			[ 'language' => 'term' | [ 'term1', 'term2', ... ] ], ...
+	 *		], ...
+	 *	]
+	 *
+	 * @return array
+	 *	[
+	 *		'type' => [
+	 *			[ 'language' => [ <textId1>, <textId2>, ... ] ], ...
+	 *		], ...
+	 *	]
+	 */
+	private function mapToTextIds( array $termsArray ) {
+		$texts = [];
+
+		array_walk_recursive( $termsArray, function ( $text ) use ( &$texts ) {
+			$texts[] = $text;
+		} );
+
+		$textIds = $this->acquireTextIds( array_unique( $texts ) );
+
+		array_walk_recursive( $termsArray, function ( &$text ) use ( $textIds ) {
+			$text = $textIds[$text];
+		} );
+
+		return $termsArray;
+	}
+
+	private function acquireTextIds( array $texts ) {
+		$existingTextRows = $this->dbr->select(
+			self::TABLE_TEXT,
+			[ 'wbx_id', 'wbx_text' ],
+			[ 'wbx_text' => $texts ]
+		);
+
+		$existingTextIds = array_reduce(
+			iterator_to_array( $existingTextRows ),
+			function ( $carry, $row ) {
+				$carry[$row->wbx_text] = $row->wbx_id;
+				return $carry;
+			},
+			[]
+		);
+
+		$nonExistingTexts = array_diff( $texts, array_keys( $existingTextIds ) );
```

Comment:
Why not abort here, if $nonExistingTexts is empty?

Reply:
it surely should